warehouse/clickhouse/clickhouse.go

```diff
@@ -0,0 +1,663 @@
+		return
+	}
+	defer misc.RemoveFilePaths(fileNames...)
+
+	txn, err := ch.Db.Begin()
+	if err != nil {
+		logger.Errorf("ch: Error while beginning a transaction in db for loading in table:%s: %v", tableName, err)
+		warehouseutils.SetTableUploadError(warehouseutils.ExportingDataFailedState, ch.Upload.ID, tableName, err, ch.DbHandle)
+		return
+	}
+
+	stmt, err := txn.Prepare(fmt.Sprintf(`INSERT INTO "%s"."%s" (%v) VALUES (%s)`, ch.Namespace, tableName, sortedColumnString, generateArgumentString("?", len(sortedColumnKeys))))
+	if err != nil {
+		logger.Errorf("ch: Error while preparing statement for  transaction in db for loading in  table:%s: %v", tableName, err)
+		warehouseutils.SetTableUploadError(warehouseutils.ExportingDataFailedState, ch.Upload.ID, tableName, err, ch.DbHandle)
+		return
+	}
+
+	for _, objectFileName := range fileNames {
+		var gzipFile *os.File
+		gzipFile, err = os.Open(objectFileName)
+		if err != nil {
+			logger.Errorf("ch: Error opening file using os.Open for file:%s while loading to table %s", objectFileName, tableName)
+			warehouseutils.SetTableUploadError(warehouseutils.ExportingDataFailedState, ch.Upload.ID, tableName, err, ch.DbHandle)
+			return
+		}
+
+		var gzipReader *gzip.Reader
+		gzipReader, err = gzip.NewReader(gzipFile)
+		if err != nil {
+			logger.Errorf("CH: Error reading file using gzip.NewReader for file:%s while loading to table %s", gzipFile, tableName)
+			rruntime.Go(func() {
+				misc.RemoveFilePaths(objectFileName)
+			})
+			warehouseutils.SetTableUploadError(warehouseutils.ExportingDataFailedState, ch.Upload.ID, tableName, err, ch.DbHandle)
+			gzipFile.Close()
+			return
+
+		}
+		csvReader := csv.NewReader(gzipReader)
+		for {
+			var record []string
+			record, err = csvReader.Read()
+			if err != nil {
+				if err == io.EOF {
+					logger.Infof("CH: File reading completed while reading csv file for loading in table:%s: %s", tableName, objectFileName)
+					break
+				} else {
+					logger.Errorf("CH: Error while reading csv file for loading in table:%s: %v", tableName, err)
+					warehouseutils.SetTableUploadError(warehouseutils.ExportingDataFailedState, ch.Upload.ID, tableName, err, ch.DbHandle)
+					txn.Rollback()
+					return
+				}
+			}
+			var recordInterface []interface{}
+			for index, value := range record {
+				columnName := sortedColumnKeys[index]
+				columnDataType := columnMap[columnName]
+				data := typeCastDataFromType(value, columnDataType)
+				recordInterface = append(recordInterface, data)
+
+			}
+
+			_, err = stmt.Exec(recordInterface...)
+			if err != nil {
+				logger.Errorf("CH: Error in exec statement for loading in table:%s: %v", tableName, err)
+				warehouseutils.SetTableUploadError(warehouseutils.ExportingDataFailedState, ch.Upload.ID, tableName, err, ch.DbHandle)
+				txn.Rollback()
+				return
+			}
+
+		}
+		gzipReader.Close()
+		gzipFile.Close()
+	}
+	if err != nil && err != io.EOF {
```

Comment:
is this required? where is this error set. all the above errors are already handled right?